src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -0,0 +1,447 @@
+    case ResponseReceived(peer, BlockHeaders(headers), _) if state.isHeadersFetcher(sender()) =>
+      log.debug("Fetched headers for blocks {} by {}", headers.map(_.number).mkString(","), sender())
+
+      val newState = state.validatedHeaders(headers) match {
+        case Left(err) =>
+          blacklistIfHandshaked(peer, err)
+          state.withoutHeadersFetcher()
+        case Right(validHeaders) =>
+          state.appendHeaders(validHeaders)
+      }
+
+      fetchBlocks(newState)
+    case RequestFailed(peer, reason) if state.isHeadersFetcher(sender()) =>
+      log.debug(s"Failed headers request from peer {} with reason {}", peer, reason)
+      blacklistIfHandshaked(peer, reason)
+      state.withoutHeadersFetcher() |> fetchBlocks
+  }
+
+  private def handleBodiesMessages(state: FetcherState): Receive = {
+    case ResponseReceived(peer, BlockBodies(bodies), _) if state.isBodiesFetcher(sender()) =>
+      log.debug("Fetched {} block bodies by {}", bodies.size, sender())
+      state.addBodies(peer, bodies) |> fetchBlocks
+    case RequestFailed(peer, reason) if state.isBodiesFetcher(sender()) =>
+      log.debug(s"Failed bodies request from peer {} with reason {}", peer, reason)
+      blacklistIfHandshaked(peer, reason)
+      state.withoutBodiesFetcher() |> fetchBlocks
+  }
+
+  private def handleStateNodeMessages(state: FetcherState): Receive = {
+    case FetchStateNode(hash) => fetchStateNode(hash, sender(), state)
+    case RetryFetchStateNode(hash, replyTo) => fetchStateNode(hash, replyTo, state)
+    case ResponseReceived(peer, NodeData(values), _) if state.isStateNodeFetcher(sender()) =>
+      log.debug("Received state node response from peer {}", peer)
+      state.stateNodeFetcher.foreach(fetcher => {
+        val validatedNode = values
+          .asRight[String]
+          .ensure(s"Empty response from peer $peer, blacklisting")(_.nonEmpty)
+          .ensure(s"Fetched node state hash doesn't match requested one, blacklisting peer")(nodes =>
+            fetcher.hash == kec256(nodes.head))
+
+        validatedNode match {
+          case Left(err) =>
+            log.debug(err)
+            blacklist(peer.id, syncConfig.blacklistDuration, err)
+            fetchStateNode(fetcher.hash, fetcher.replyTo, state)
+          case Right(node) =>
+            fetcher.replyTo ! FetchedStateNode(NodeData(node))
+            context become started(state.withoutStateNodeFetcher())
+        }
+      })
+    case RequestFailed(peer, reason) if state.isStateNodeFetcher(sender()) =>
+      log.debug(s"Failed state node request from peer {} with reason {}", peer, reason)
+      blacklistIfHandshaked(peer, reason)
+      state.stateNodeFetcher.foreach(fetcher => {
+        fetchStateNode(fetcher.hash, fetcher.replyTo, state.withoutHeadersFetcher())
+      })
+  }
+
+  private def handleNewBlockMessages(state: FetcherState): Receive = {
+    case MessageFromPeer(NewBlockHashes(_), _) =>
```

Comment:
Why such handling of new block hashes ? In principle we should request header for that hash, and then body and then do import.

Reply:
I remember there was something weird about handling this message and it appeared to me better to just schedule next fetch instead of handling it directly. I will check it in detail and provide more detailed response.

Reply:
The issue is that request for headers still has to be made. So instead of fetching headers for that specific hashes (while it's possible that there are some missing headers in between) is worse option that just doing a request for next bunch of headers and then bodies.